### ISSUE TITLE
Unclear error message when a numpy array is not C_CONTIGUOUS

### ISSUE THREAD

**Issue by sm-Fifteen** [ijl/orjson#100]
Attempting to serialize a numpy array that does not happen to have an underlying C_CONTIGUOUS representation will have orjson throw the same `orjson.JSONEncodeError: Type is not JSON serializable: numpy.ndarray` as it does when numpy serialization is disabled, with no clear indication that this is the real cause.

```py
import numpy
import orjson

f_array = numpy.array([[1,2,3,4,5],[6,7,8,9,0]], order='F')
c_array = numpy.ascontiguousarray(f_array)

c_json = orjson.dumps(c_array, option=orjson.OPT_SERIALIZE_NUMPY) # Works fine
f_json = orjson.dumps(f_array, option=orjson.OPT_SERIALIZE_NUMPY) # orjson.JSONEncodeError: Type is not JSON serializable: numpy.ndarray

print(c_json)
print(f_json)
```

While this example might seem contrived, I just encountered this error while trying to serialize select lines from a pandas dataframe, which generally doesn't let you specify what array order the internal arrays are stored in. Based on the error message, one would probably just assume numpy serialization to be buggy or broken.

**Comment by ijl** [comment]
The error message can be specific to this. It can further distinguish between `PyArrayError::NotContiguous` and `PyArrayError::UnsupportedDataType` by passing a new optional parameter to `DefaultSerializer`.

There is documentation in the numpy section that this is intended behavior. It could include a full-up example of both fallthrough cases.

**Comment by sm-Fifteen** [comment]
> There is documentation in the numpy section that this is intended behavior. It could include a full-up example of both fallthrough cases.

That's how I figured what the actual problem was. The error message wasn't all that helpful in that regard, but I remembered the documentation making mention of that requirement and my data structure turned out to only contain one numby array, so I at least didn't spend too long guessing.

> The error message can be specific to this. It can further distinguish between `PyArrayError::NotContiguous` and `PyArrayError::UnsupportedDataType` by passing a new optional parameter to `DefaultSerializer`.

Hmm, yeah, ok, I see it, you're trying to passthrough the bad numpy arrays to `default` instead of outright rejecting them...

https://github.com/ijl/orjson/blob/a95eb5635d5a1b4af864c3e6767540651de6d2f2/src/encode.rs#L426-L439

So that would mean that error message is actually about `default` not supporting `numpy.ndarray`, rather than the encoder itself?
